### PR TITLE
Read_NPX to support Parquet input (OA-375)

### PR DESCRIPTION
**Pull request by kathy-nevola**
# Title: Support for parquet file type
**Problem:**Parquet filetype not currently supported in OA

**Solution:** Add function, support, and unit tests to support

**Key Features:**

* read_npx_parquet function
* example parquet file
* unit tests which parallel existing file types

## Checklist
- [X] Make sure you are requesting to **pull a feature/bugfix branch** (right side). This should not be main or develop.
- [X] Make sure you are make a pull request against either **main or develop** (left side). (Requesting to main should be reserved for bugfixes and new releases)
- [X] Add or update unit tests (if applicable)
- [X] Check your code with any unit tests (Run devtools::check() locally)
- [NA] Add neccessary documentation (if applicable)

## Type of changes

What type of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue) (link the issue on the right)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Documentation Update 
- [ ] Other (explain)

## Further comments
Code was primarily written by @klevdiamanti and unit tests added by @kathy-nevola

### CODE REVIEW

File: OlinkAnalyze/R/read_flex.R [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+                                               "Plate LOD",
+                                               "LLOQ",
+                                               "ULOQ",
+                                               "LOD", 
+                                               collapse = "|"))) |>  # remove end meta data
+    tidyr::pivot_longer(cols = tidyselect::starts_with("OID"),
+                        names_to = "OlinkID",
+                        values_to = ifelse(is_npx_data, "NPX", "Quantified_value")) |> 
+    dplyr::left_join(metadata, by = c("OlinkID")) |> 
+    dplyr::left_join(meta_data_per_assay, multiple = "all") |>
```
Comment (Orbmac) [inline OlinkAnalyze/R/read_flex.R:166]: This line 
"dplyr::left_join(meta_data_per_assay, multiple = "all")" 
will print out "Joining with `by = join_by(OlinkID)`" which might be confusing when reading in data.

File: OlinkAnalyze/tests/testthat/test-Read_NPX_data.R [hunk 2]
```diff
@@ -1,142 +1,156 @@
...
+  expect_error(read_NPX(filename = zip_npx_file_fail_1), "Checksum of NPX file does not match the one from \"MD5_checksum.txt\"! Loss of data?")
+
+  zip_npx_file_fail_2 <- system.file("extdata", "Example_NPX_Data_empty.zip", package = "OlinkAnalyze", mustWork = TRUE)
+  expect_error(read_NPX(filename = zip_npx_file_fail_2), "The compressed file does not contain a valid NPX file. Expecting: \"README.txt\", \"MD5_checksum.txt\" or \"checksum_sha256.txt\" and the NPX file.")
+
+  zip_npx_file_success <- system.file("extdata", "Example_NPX_Data_3K.zip", package = "OlinkAnalyze", mustWork = TRUE)
+  df_2 <- read_NPX(filename = zip_npx_file_success)
+
+  zip_npx_file_success_sha <- system.file("extdata", "Example_NPX_Data_sha256.zip", package = "OlinkAnalyze", mustWork = TRUE)
+  expect_snapshot(read_NPX(filename = zip_npx_file_success_sha))
```
Comment (Orbmac) [inline OlinkAnalyze/tests/testthat/test-Read_NPX_data.R:39]: expect_snapshot(read_NPX(filename = zip_npx_file_success_sha))

throws a fail for me. Not sure really what this is suppose to do. It compares to a previous run of it?

**Review by Orbmac — COMMENTED** [review]
Just a few comments, but overall everything seem to be working.

File: OlinkAnalyze/R/Read_NPX_data.R [hunk 3]
```diff
@@ -39,171 +39,131 @@
...
-                   "header_ext_v1"   = c(header_ext_standard,
-                                         "INC_Warning",
-                                         "AMP_Warning",
-                                         "Count_Warning"),
-                   "header_ext_v2"   = c(header_ext_standard,
-                                         "ExploreVersion"))
-
-  header_match <-  header_v |>
+
+  header_v <- list(
```
[the remark this answers is not in the document]
Comment (klevdiamanti) [reply OlinkAnalyze/R/Read_NPX_data.R:87]: This has been updated continuously while explore deliverable files have changed. For the earlier changes there is limited records, and newer ones have been documented. This is expected to be handled better later this year.

File: OlinkAnalyze/R/Read_NPX_data.R [hunk 4]
```diff
@@ -213,99 +173,133 @@ read_NPX_explore <- function(filename) {
...
+                                       col_names = FALSE,
+                                       .name_repair = "minimal")
+
+  if (any(stringr::str_detect(c(panel_name, panel_name_long), pattern = "Target 48"))){
+
+    isTarget48 <- TRUE
+    isTarget96 <- FALSE
+    isFlex <- FALSE
+
+  } else if (any(stringr::str_detect(c(panel_name, panel_name_long), pattern = "Target 96"))){
```
[the remark this answers is not in the document]
Comment (kathy-nevola) [reply OlinkAnalyze/R/Read_NPX_data.R:271]: Good point! All the newer versions of Target 96 files contain Target 96 in the panel  name, but I dont know that older versions did. May be worth updating npx_data2 to fit the newest naming convention.

File: OlinkAnalyze/tests/testthat/test-Read_NPX_data.R (hunk 2 above)
Comment (kathy-nevola) [reply OlinkAnalyze/tests/testthat/test-Read_NPX_data.R:39]: @Orbmac Interesting, it doesnt fail locally for me.  
expect_snapshot compares the output of reading the file to the Read_NPX_data.md in tests/testthat/_snaps. 

If you run 
`zip_npx_file_success_sha <- system.file("extdata", "Example_NPX_Data_sha256.zip", package = "OlinkAnalyze", mustWork = TRUE)`
and then 
`read_NPX(filename = zip_npx_file_success_sha)`

are there any differences in the console output compared to the Read_NPX_data.md file?

Comment (Orbmac) [reply OlinkAnalyze/tests/testthat/test-Read_NPX_data.R:39]: I assume its something locally on my computer then if it works for you. The values was the same, the only difference I could see was that it said "Missing~1" while for me it said ""Missing..1" (similar for other long col names).

**Review by brittalot — COMMENTED** [review]
Tested read_nxp_csv and read_npx_parquet without any errors in Shiny Concordance app. No further comments.

**Pull request merged**